### ISSUE TITLE
Tensorflow Lite: Using output index causes segfault. 

### ISSUE THREAD

**Issue by cyrusbehr** [tensorflow/tensorflow#53951]
I am using Tensorflow Lite C++ interface for running inference, commit hash `040585c0f25681b399c9087b53c982959bcca44f` (HEAD of master branch at the time of posting this). 
I am using a model with 2 inputs, and a single output (array). 

I am trying to run the following code as a sanity check: 

```
#include <iostream>
#include <fstream>
#include <vector>
#include "tensorflow/lite/interpreter.h"
#include "tensorflow/lite/kernels/register.h"
#include "tensorflow/lite/model.h"

#include <opencv2/opencv.hpp>

using namespace tflite;

int main() {
    std::ifstream modelFile ("../models/liveness.tflite", std::ios::binary | std::ios::ate);
    std::streamsize size = modelFile.tellg();
    modelFile.seekg(0, std::ios::beg);

    std::vector<char> buffer(size);

    if (!modelFile.read(buffer.data(), size)) {
        throw std::runtime_error("There was a issue reading the model file into memory");
    }

    StderrReporter errorReporter;
    auto model = FlatBufferModel::BuildFromBuffer(buffer.data(), size, &errorReporter);
    if (model == nullptr) {
        throw std::runtime_error("Error creating builder from model file");
    }

    tflite::ops::builtin::BuiltinOpResolver resolver;

    std::unique_ptr<Interpreter> interpreter;

    if (InterpreterBuilder(*model, resolver)(&interpreter) != kTfLiteOk) {
        throw std::runtime_error("Unable to create interpreter");
        // Return failure.
    }

    auto inputs = interpreter->inputs();
    for (const auto input: inputs) {
        std::cout << "Input: " << input << std::endl;
        std::cout << "Input name: " << interpreter->GetInputName(input) << std::endl;
    }

    auto outputs = interpreter->outputs();
    for (const auto output: outputs) {
        std::cout << "Output: " << output << std::endl;
        std::cout << "Output name: " << interpreter->GetOutputName(output) << std::endl;
    }

    return 0;
}
```

When I run the above, it prints:
```
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
Input: 199
Input name: batch_normalization/batchnorm/add_1
Input: 200
Input name: batch_normalization/batchnorm/add_1
Output: 193
Segmentation fault (core dumped)

```
As can be seen, the call to ` std::cout << "Output name: " << interpreter->GetOutputName(output) << std::endl;` causes a segfault. 

If I instead change it to ` std::cout << "Output name: " << interpreter->GetOutputName(0) << std::endl;` it prints `Output name: dense/Softmax` 

Why is the first version calling a segfault? 

As a sanity check, I use python to run this snippet of code with the same model:
```
    output_details = interpreter.get_output_details()

    print(output_details)

```

This prints:

```
[{'name': 'dense/Softmax', 'index': 193, 'shape': array([1, 2], dtype=int32), 'shape_signature': array([1, 2], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}]

```

The reason I even ask this question is because I expect to be able to obtain the inference results by calling:
```
float* out = interpreter->typed_output_tensor<float>(interpreter->outputs()[0]);
```

However, this too causes a segmentation fault. 

Running `float* out = interpreter->typed_output_tensor<float>(0);` does give me the correct output.

**Comment by haozha111** [comment]
hi,

the `interpreter->GetOutputName(i)` API expects the index to between 0 and the outputs.size() - 1, it's not the tensor index. When you use `output` which returned from the `interpreter->outputs()`, this is the output tensor index in the graph, so it might be out of bound.

**Comment by cyrusbehr** [comment]
Makes sense, thank you!
